plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -545,6 +555,159 @@ public ColumnHandle getDeleteRowIdColumnHandle(ConnectorSession session, Connect
+        Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
+
+        verify(transaction == null, "transaction already set");
+        transaction = icebergTable.newTransaction();
+
+        return new BeginTableExecuteResult<>(
+                executeHandle.withProcedureHandle(optimizeHandle.withSnapshotId(Optional.ofNullable(icebergTable.currentSnapshot()).map(Snapshot::snapshotId))),
+                table.forOptimize(true, optimizeHandle.getMaxScannedFileSize()));
+    }
+
+    @Override
+    public void finishTableExecute(ConnectorSession session, ConnectorTableExecuteHandle tableExecuteHandle, Collection<Slice> fragments, List<Object> splitSourceInfo)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        switch (executeHandle.getProcedureId()) {
+            case OPTIMIZE:
+                finishOptimize(executeHandle, fragments, splitSourceInfo);
+                return;
+        }
+        throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+    }
+
+    private void finishOptimize(IcebergTableExecuteHandle executeHandle, Collection<Slice> fragments, List<Object> splitSourceInfo)
+    {
+        IcebergOptimizeHandle optimizeHandle = (IcebergOptimizeHandle) executeHandle.getProcedureHandle();
+        Table icebergTable = transaction.table();
+
+        // paths to be deleted
+        Set<DataFile> scannedFiles = splitSourceInfo.stream()
+                .map(DataFile.class::cast)
+                .collect(toImmutableSet());
+
+        List<CommitTaskData> commitTasks = fragments.stream()
+                .map(slice -> commitTaskCodec.fromJson(slice.getBytes()))
+                .collect(toImmutableList());
+
+        Type[] partitionColumnTypes = icebergTable.spec().fields().stream()
+                .map(field -> field.transform().getResultType(
+                        icebergTable.schema().findType(field.sourceId())))
+                .toArray(Type[]::new);
+
+        Set<DataFile> newFiles = new HashSet<>();
+        for (CommitTaskData task : commitTasks) {
+            DataFiles.Builder builder = DataFiles.builder(icebergTable.spec())
+                    .withPath(task.getPath())
+                    .withFileSizeInBytes(task.getFileSizeInBytes())
+                    .withFormat(optimizeHandle.getFileFormat())
+                    .withMetrics(task.getMetrics().metrics());
+
+            if (!icebergTable.spec().fields().isEmpty()) {
+                String partitionDataJson = task.getPartitionDataJson()
+                        .orElseThrow(() -> new VerifyException("No partition data for partitioned table"));
+                builder.withPartition(PartitionData.fromJson(partitionDataJson, partitionColumnTypes));
+            }
+
+            newFiles.add(builder.build());
+        }
+
+        if (scannedFiles.isEmpty() && newFiles.isEmpty()) {
```

Comment:
assert we should not ever get one empty and other not? Feels like a bug situation.

Reply:
Scanned file list may be non empty, but resulting data may be empty, if input files were empty.